src/storage/tde.h

```diff
@@ -101,6 +101,15 @@ typedef struct tde_mk_file_item
 #endif /* CS_MODE */
 #endif /* UNSTABLE_TDE_FOR_REPLICATION_LOG */
 
+#if defined(CS_MODE)
+typedef struct dblink_cipher
+{
+  bool is_loaded;
+  unsigned char master_key[TDE_DATA_KEY_LENGTH];
+} DBLINK_CHPHER_KEY;
+extern DBLINK_CHPHER_KEY dblink_Cipher_key;
+#endif
```

Comment:
This structure doesn't have to be here. It is not related to TDE itself. Also, I think this file should be dealt with as a server's module (Only SERVER_MODE and SA_MODE can include this file).
How about adding authors (@wrlawodms) as a reviewer?

Reply:
@wrlawodms
     You can only look at the TDE related part.
    The relevant sources are the files below.
`    tde.h, tde.c, network_interface_cl.c, crypt_opfunc.c`

   You can find it in the source with the "dblink_Cipher_key" variable name.

Reply:
I took a look at the code. To begin with, I'd like to ask you for the overall architecture for the encryption.
1. Please elaborate on the data to be encrypted. Are they decrypted and encrypted only in the same database instance? How about their lifecycle? They will be gone after the server shutdown?
2. Which value did you choose as NONCE? they must be unique for every message.
3. Which values can be used as the master key? the key must not be hard-corded or retrievable.
- BTW, I don't think the name "master key" is good. I used it in TDE because of 2-level encryption. You seem that you need only one key.

Reply:
And, generally, I don't recommend using constants and interfaces for TDE.  They are defined for the TDE's requirements and designs. They can be changed when we revise the TDE design or deal with issues on TDE. Please decouple it.

Reply:
@wrlawodms
We are adding a new feature called DBLINK.
A catalog called db_server is added to support this feature.
A password is included to store access information to the remote DB in this catalog, and this password must be able to be decrypted.
It uses the AES algorithm, and for this purpose, it was written by referring to the processing code of the TDE-related source.
There's more, but I'll leave out the parts that are beyond the scope of what wrlawodms will review.
1. The string itself is encrypted with the password of the remote DB and stored as a column value in the db_server catalog. Since AES comes in binary, it is saved as a base64 readable string.

2. The nonce is 16 bytes in size, same as the tde source,
The value is filled in using the code below.
`memset(evp_cipher.nonce, 0x07, sizeof (evp_cipher.nonce));`

3. Obtain the tde_Cipher.data_keys.log_key value generated from the tde-related code and use it as it is. What is the master key here?
It means the value passed to the key parameter when calling the EVP_EncryptInit_ex() function.



Here, we do not use the TDE function, but just use "tde_Cipher.data_keys.log_key" that is used for encryption processing in the TDE function. This is because there is an expectation that, without exposing the master key, it will always be possible to obtain a value that does not change in the same DB.

Reply:
Thank you for answering!
2. Basically, NONCE (Number used ONCE) has to be different for each data encrypted, but you are using the hard-coded number, 0x070707. Is it ok?
3. "Master key" is a concept for TDE key management, which is irrelevant to general encryption.

One more question: TDE assumes the master key for TDE could be lost, which in turn means you can't access the `log_key`. In this case, DB just works without encrypted data. If you use `log_key` to encrypt the password, the DB link feature will be dependant on the TDE Key management policy. Is it on purpose?

I'll leave other reviews on each code line.

Reply:
@wrlawodms
    I refer to the code below inside the tde_dk_nonce() function
```
  case TDE_DATA_KEY_TYPE_LOG:
       memset(dk_nonce, 2, TDE_DK_NONCE_LENGTH);
```
I don't know what the exact role of the nonce is, but I think that using the nonce value as a fixed value will not cause any problems.

If the TDE key (<dbname>_keys file) is lost, the TDE function will be limited and DBLINK's encryption processing will not be possible.
However, other functions except dBLINK can be used as they are now.

What I am concerned about is this.
It should be guaranteed that the value of tde_Cipher.data_keys.log_key cannot be changed in the DB that has already been created (except for loss).

Reply:
> I don't know what the exact role of the nonce is, but I think that using the nonce value as a fixed value will not cause any problems.

In the code you referred to, you can see each data key is encrypted with a difference NONCE. Please refer to `tde_encrpyt_data_page()` which use a different NONCE for each encrypted data. In your case, I think NONCE has to be different per each data (maybe per each user password). In my opinion, if they are stored in a catalog class, you can use each OID of record as the NONCE. From what I know, this is a security requirement and if NONCE of two encrypted data is the same, an attacker can guess whether two data are the same or not. 

> If the TDE key (_keys file) is lost, the TDE function will be limited and DBLINK's encryption processing will not be possible.
> However, other functions except dBLINK can be used as they are now.

I don't really know how dBLINK works. If it can work without the encrypted data (maybe the user password for a remote server), it doesn't matter. 

> What I am concerned about is this.
> It should be guaranteed that the value of tde_Cipher.data_keys.log_key cannot be changed in the DB that has already been created (except for loss).

Yes, the data key is guaranteed not to be changed.

Reply:
Oh, It doesn't seem enough to use the record OID for a Nonce for it. With this, if a user changed the password with the same word, an attacker can know it. I think you should figure out another unique nonce for each data if you want to meet the requirement strictly.